File: app/src/main/java/com/ad/whatsappclone/adapter/ChatMessageAdapter.java
package com.ad.whatsappclone.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ad.whatsappclone.R;
import com.ad.whatsappclone.models.ChatMessages;
import com.google.firebase.auth.FirebaseAuth;

import java.util.ArrayList;
import java.util.List;

public class ChatMessageAdapter extends RecyclerView.Adapter {
    private static final int SENDER_VIEW_TYPE = 0;
    private static final int RECEIVER_VIEW_TYPE = 1;
    List<ChatMessages> chatMessages = new ArrayList<>();
    Context context;


    public ChatMessageAdapter(List<ChatMessages> chatMessages, Context context) {
        this.chatMessages = chatMessages;
        this.context = context;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == SENDER_VIEW_TYPE) {
            View view = LayoutInflater.from(context).inflate(R.layout.recycle_sender, parent, false);
            return new SenderViewHolder(view);
        } else {
            View view = LayoutInflater.from(context).inflate(R.layout.recycle_receiver, parent, false);
            return new ReceiverViewHolder(view);
        }

    }

    @Override
    public int getItemViewType(int position) {
        if (chatMessages.get(position).getMessageId().equals(FirebaseAuth.getInstance().getUid())) {
            return SENDER_VIEW_TYPE;
        } else {
            return RECEIVER_VIEW_TYPE;
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        ChatMessages message = chatMessages.get(position);
        if (holder.getClass() == SenderViewHolder.class) {
            ((SenderViewHolder) holder).senderMessage.setText(message.getMessage());
        } else {
            ((ReceiverViewHolder) holder).receiverMessage.setText(message.getMessage());
        }
    }

    @Override
    public int getItemCount() {
        return chatMessages.size();
    }

    public class ReceiverViewHolder extends RecyclerView.ViewHolder {
        TextView receiverMessage, receiverTime;

        public ReceiverViewHolder(@NonNull View itemView) {
            super(itemView);
            receiverMessage = itemView.findViewById(R.id.txtReceiverMsg);
            receiverTime = itemView.findViewById(R.id.txtReceiverTime);
        }
    }

    public class SenderViewHolder extends RecyclerView.ViewHolder {
        TextView senderMessage, senderTime;

        public SenderViewHolder(@NonNull View itemView) {
            super(itemView);
            senderMessage = itemView.findViewById(R.id.txtSenderMsg);
            senderTime = itemView.findViewById(R.id.txtSenderTime);
        }
    }
}

File: app/src/main/java/com/ad/whatsappclone/ui/home/adapter/ChatMessageAdapter.kt
package com.ad.whatsappclone.ui.home.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.ad.whatsappclone.R
import com.ad.whatsappclone.models.ChatMessages
import com.google.firebase.auth.FirebaseAuth

class ChatMessageAdapter(private val chatMessages: MutableList<ChatMessages?>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == SENDER_VIEW_TYPE) {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.recycle_sender, parent, false)
            SenderViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.recycle_receiver, parent, false)
            ReceiverViewHolder(view)
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (chatMessages[position]?.messageId == FirebaseAuth.getInstance().uid) {
            SENDER_VIEW_TYPE
        } else {
            RECEIVER_VIEW_TYPE
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val message = chatMessages[position]

        if (holder is SenderViewHolder)
            holder.senderMessage.text = message?.message
        else if (holder is ReceiverViewHolder)
            holder.receiverMessage.text = message?.message

    }

    override fun getItemCount() = chatMessages.size


    inner class ReceiverViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var receiverMessage: TextView
        private var receiverTime: TextView

        init {
            receiverMessage = itemView.findViewById(R.id.txtReceiverMsg)
            receiverTime = itemView.findViewById(R.id.txtReceiverTime)
        }
    }

    inner class SenderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var senderMessage: TextView
        private var senderTime: TextView

        init {
            senderMessage = itemView.findViewById(R.id.txtSenderMsg)
            senderTime = itemView.findViewById(R.id.txtSenderTime)
        }
    }

    companion object {
        private const val SENDER_VIEW_TYPE = 0
        private const val RECEIVER_VIEW_TYPE = 1
    }
}